nixos/modules/services/web-apps/librespeed.nix

```diff
@@ -0,0 +1,446 @@
+              };
+
+              server = lib.mkOption {
+                description = "URL to the server. You may use `//` instead of `http://` or `https://`.";
+                type = lib.types.nonEmptyStr;
+              };
+
+              dlURL = lib.mkOption {
+                description = ''
+                  URL path to download test on this server.
+                  Append `.php` to the default value if the server uses the php implementation.
+                '';
+                default = "backend/garbage";
+                type = lib.types.nonEmptyStr;
+              };
+
+              ulURL = lib.mkOption {
+                description = ''
+                  URL path to upload test on this server.
+                  Append `.php` to the default value if the server uses the php implementation.
+                '';
+                default = "backend/empty";
+                type = lib.types.nonEmptyStr;
+              };
+
+              pingURL = lib.mkOption {
+                description = ''
+                  URL path to latency/jitter test on this server.
+                  Append `.php` to the default value if the server uses the php implementation.
+                '';
+                default = "backend/empty";
+                type = lib.types.nonEmptyStr;
+              };
+
+              getIpURL = lib.mkOption {
+                description = ''
+                  URL path to IP lookup on this server.
+                  Append `.php` to the default value if the server uses the php implementation.
+                '';
+                default = "backend/getIP";
+                type = lib.types.nonEmptyStr;
+              };
+            };
+          }
+        );
+      };
+    };
+  };
+  config = lib.mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = cfg.frontend.useNginx -> cfg.domain != null;
+        message = "${opt.frontend.useNginx} requires ${opt.domain} to be set.";
+      }
+    ];
+
+    security.acme.certs = lib.mkIf (cfg.useACMEHost != null) {
+      ${cfg.useACMEHost}.reloadServices = [ "librespeed.service" ];
+    };
+
+    services = {
+      librespeed = {
+        frontend.servers = lib.mkIf (cfg.frontend.enable && (cfg.domain != null)) [
+          {
+            name = cfg.domain;
+            server = "//${cfg.domain}";
+          }
+        ];
+        settings = lib.mapAttrs (n: v: lib.mkDefault v) opt.settings.default;
+      };
+
+      nginx.virtualHosts = lib.mkIf (cfg.frontend.enable && cfg.frontend.useNginx) {
+        ${cfg.domain} = {
+          forceSSL = true;
+          locations = {
+            "/".root = librespeedAssets;
+            "= /servers.json".return = "200 '${builtins.toJSON cfg.frontend.servers}'";
+            "/backend/" = {
+              proxyPass = "http://127.0.0.1:${toString cfg.settings.listen_port}/backend/";
+              extraConfig = # nginx
+                ''
+                  brotli off;
```

Comment:
This causes nginx to crash for me with:

```
nginx: [emerg] unknown directive "brotli" in /nix/store/gh27849ccnmai77vd20iyp6hsb4fzmaz-nginx.conf:715
```

This problem goes away if I enable `services.nginx.recommendedBrotliSettings`.

Reply:
That is indeed a problem 😅 We actually need to turn this off, as otherwise the compression ratio throws off the upload speed. For example, we saw a 10x over our real upload speed.

Maybe we can check services.nginx.additionalModules for brotli and zstd and then turn them off? I think about it.